### ISSUE TITLE
Crash loop causes many windows to spawn in quick succession

### ISSUE THREAD

**Issue by axelson** [boydm/scenic#178]
### Checklist

- [x] Check other issues and make sure that it is not reported yet.

### Versions and Environment

Elixir:

```bash
# elixir -v
Elixir 1.9.1 (compiled with Erlang/OTP 22
```

Erlang:

```bash
# erl -v
rlang/OTP 22 [erts-10.4.4] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [hipe]
```

Scenic:

```text
# mix deps | grep scenic
* scenic 0.10.2 (Hex package) (mix)
  locked at 0.10.2 (scenic) f4f0a00e
* scenic_driver_glfw 0.10.1 (Hex package) (mix)
  locked at 0.10.1 (scenic_driver_glfw) df7a11bc
```

OS:

```text
Arch Linux (but applies to other OS as well)
```

### Steps to reproduce

1. `git clone https://github.com/axelson/scenic_restart_crash`
2. `mix deps.get`
3. `mix scenic.run`

### Expected Behavior

Scenic to crash a few times and then give up with the application exiting

### Actual Behavior

Many windows spawned and died, resulting in `Maximum number of clients reached` being printed multiple times and a total of around 4000 lines of output (most of it stack traces). The windows spawn (stealing the focus of the mouse) and then die so fast that it's nearly impossible to hit `ctrl-c` and kill the scenic application.

Note: When this happens scenic tries to launch a new window 256 times in under a second which causes the OS to become sluggish and unresponsive and all the windows don't fully get created and disappear until around 10 seconds or so. It's 256 times due to the current supervision hierarchy which has 4 supervisors allowing 3 restarts (so 4 starts/crashes) which is `4^4 = 256`.

#### Stack Trace

<details>

```bash
07:23:49.391 [error] GenServer :main_viewport terminating
** (UndefinedFunctionError) function ScenicRestartCrash.Scene.Home.start_dynamic_scene/4 is undefined (module ScenicRestartCrash.Scene.Home is not available)
    ScenicRestartCrash.Scene.Home.start_dynamic_scene(#PID<0.2017.0>, nil, nil, [vp_dynamic_root: #PID<0.2014.0>, viewport: #PID<0.2014.0>, styles: %{}])
    (scenic) lib/scenic/view_port.ex:532: Scenic.ViewPort.handle_cast/2
    (stdlib) gen_server.erl:637: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:711: :gen_server.handle_msg/6
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Last message: {:"$gen_cast", {:set_root, {ScenicRestartCrash.Scene.Home, nil}, nil}}
State: %{default_scene: {ScenicRestartCrash.Scene.Home, nil}, default_scene_activation: nil, driver_registry: %{}, drivers: [], dynamic_root_pid: nil, dynamic_supervisor: #PID<0.2017.0>, hover_primitve: nil, input_captures: %{}, master_graph: %{0 => %{data: {Scenic.Primitive.Group, [1]}, transforms: %{}}, 1 => %{data: {Scenic.Primitive.SceneRef, nil}}}, master_graph_key: {:graph, #Reference<0.1457681126.2594177025.129008>, nil}, master_styles: %{}, master_transforms: %{}, max_depth: 256, on_close: :stop_system, root_config: nil, root_graph_key: nil, root_scene_pid: nil, size: {700, 600}, supervisor: #PID<0.2005.0>}
Maximum number of clients reached
```

</details>

### Additional Comments

The specific crash isn't as interesting to me (and could be relatively easily fixed), but I would like to find a way to categorically not end up in this situation either through a change in the supervision hierarchy or possibly custom supervision.

Some half-baked ideas for solutions:
1. Add a delay on restarting the Viewport if it crashes
  a. Pro: This would give the user time to manually stop the scenic process
  b. Con: Depending on the length of the delay, some of the supervisors will constantly crash because the lower-level supervisors are not crashing often enough in the default 5 second window to crash the higher-level supervisor.
2. Reduce the depth of the supervision hierarchy by removing supervisors or turning down the max number of crashes that they allow and possibly increasing the `max_seconds` configuration
a. Pro: This will mitigate the issue
b. Con: The issue will still exist, just in a less annoying format
3. Use some sort of custom supervision strategy
a. Pro: It is probably possible to prevent the issue
b. Con: More code complexity and likely more difficulty understanding the system as a whole

I am not sure which direction would be best at this point. I kind of like option 2, although depending on the necessary complexity option 3 might be best. Perhaps after enough failed restarts in a row the viewport's supervisor could somehow mark itself as "unrecoverable" and would no longer try to restart it's children.

**Comment by BWheatie** [comment]
👍  to this. Has happened a few times on MacOS and ends up crashing my computer.

**Comment by boydm** [comment]
Pretty sure this is finally fixed in v0.11. Reopen if you see it again please.

v0.11 will be merged into master soon
